src/unit_tests/method/method_order_test.cpp

```diff
@@ -0,0 +1,181 @@
+*   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE     *
+*   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER          *
+*   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,   *
+*   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE   *
+*   SOFTWARE.                                                                       *
+*                                                                                   *
+*************************************************************************************/
+
+#include <rttr/registration>
+#include <catch/catch.hpp>
+#include <string>
+
+using namespace rttr;
+
+struct method_order_test_base
+{
+    method_order_test_base() = default ;
+    virtual ~method_order_test_base() = default ;
+
+    std::string whoami() { return "I am a base method" ; }
+    virtual std::string vwhoami() { return "I am a virtual base method" ; }
+
+    RTTR_ENABLE()
+
+};
+
+
+// RTTR_ENABLE incompatible with clang
+#define RTTR_ENABLE_OVERRIDE(...) \
+public:\
+    RTTR_INLINE rttr::type get_type() const override { return rttr::detail::get_type_from_instance(this); }  \
+    RTTR_INLINE void* get_ptr() override { return reinterpret_cast<void*>(this); } \
+    RTTR_INLINE rttr::detail::derived_info get_derived_info() override { return {reinterpret_cast<void*>(this), rttr::detail::get_type_from_instance(this)}; } \
+    using base_class_list = TYPE_LIST(__VA_ARGS__);
+
+
+struct method_order_test_derived : method_order_test_base
+{
+    method_order_test_derived() = default ;
+    ~method_order_test_derived() override = default ;
+
+    std::string whoami() { return "I am a non-virtual derived method with same name" ; }
+    std::string vwhoami() override { return "I am a virtual derived method with same name" ; }
+
+    RTTR_ENABLE_OVERRIDE(method_order_test_base)
+};
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+RTTR_REGISTRATION
+{
+    registration::class_<method_order_test_base>("method_order_test_base")
+        .method("whoami", &method_order_test_base::whoami)
+        .method("vwhoami", &method_order_test_base::vwhoami)
+        ;
+    registration::class_<method_order_test_derived>("method_order_test_derived")
+        .method("whoami", &method_order_test_derived::whoami)
+        .method("vwhoami", &method_order_test_derived::vwhoami)
+        ;
+}
+
+// approach 1 looks up the "method instance" without considering argument type and invoke the registered function()
+template<class TT>
+std::string rttr_invoke_approach1(
+        TT& inst,
+        const std::string& meth_name) {
+
+    const auto& inst_t = type::get<TT>();
+
+    // find method
+    const auto& meth = inst_t.get_method(meth_name);
+    REQUIRE(meth.is_valid() == true);
+
+    // invoke via method
+    variant&& iam_var = meth.invoke(inst);
```

Comment:
Why `&&` ?
Should be enough:
`auto iam_var = meth.invoke(inst);`

Reply:
References are better than variables because they ensure no copies are made.

Best
1.  const ref& - ensures reference cannot be modified.  When you see this you know what you have.
2.  ref&& - universal reference allows modification as lvalue
3.  variable - requires a copy
Worst

Reply:
Done - const auto& works

Reply:
Please see here:
https://stackoverflow.com/questions/4986673/c11-rvalues-and-move-semantics-confusion-return-statement
The variant has a move-ctor. So just use `auto` and you are fine.